### PR TITLE
OpenMP initial integration

### PR DESCRIPTION
**Pull request by dsroche**
This is my initial stab at OpenMP for FLINT. Here's what I've done specifically:

+ Added configure options to enable/disable openmp support. It's currently disabled by default.
+ Integrate with existing code, mostly adding `#pragma omp threadprivate()` declarations wherever something was declared using TLS.
+ Added a new function `flint_parallel_cleanup()` which calls the cleanup functions to clear caches etc. *unless* you're in the master thread.
+ A new test in `test/t-omp-cleanup.c` checks OpenMP support and that the cleanup function is working as promised.
+ There's a new flag in `flint.h`, `FLINT_PREFER_OMP, which is 1 if OpenMP is enabled, or if pthreads is *disabled*. (The logic being, OpenMP code will still run fine if the pragmas are ignored, but if you don't have pthreads that code will simply not work.)
+ The `fmpz_poly_taylor_shift_multi_mod` function now checks `FLINT_PREFER_OMP` to decide whether to use the pthreads implementation (already in Flint) or the OpenMP version that I made up.

This is all still undocumented, mostly because I wasn't sure where to add things in the documentation! Perhaps there should be a new section in `flint-manual.tex` on multithreading?

As for actually using OpenMP in the Flint library, there are certainly many opportunities! I have a few recommendations for all of us:
+ Separate the parallel construct  `omp parallel` from the work-sharing constructs like `omp for`, `omp sections`, etc. This way, the cleanup can be called once per thread, no matter how the work is split up.
+ Don't just throw omp pragmas on every loop you see. There is a penalty (maybe a few dozen milliseconds or more) to spinning up threads and cleaning them up, so things should only be run in parallel when there is a reasonable expectation that with at least two cores, the execution will actually be faster.
+ Regarding the previous point, I think in many cases it would be advisable to have a separate `_omp` version of a function that is only called when the problem size is large enough, or if the user explicitly wants it.
+ Nesting OpenMP parallel sections within each other, or with pthreads, is not a good idea and we should avoid it. This also means not calling external libraries that might be threaded (such as Atlas BLAS) from within a parallel section of Flint.
+ Stick to the OpenMP 2.0 standard if we can, as this should be most widely supported. Also prefer if at all possible to use the `#pragma omp` syntax rather than the library calls from `omp.h`. This will ensure the code runs correctly even when OpenMP support is not enabled.

### CODE REVIEW

**Comment by wbhart** [comment]
Sorry it's taken me so long to look at this. This whole year has been so busy for me.

PREFER_OMP seems odd compared to WANT_OMP, but I understand why you called it that. It's not enough of an objection for me to not merge this. Thanks very much for doing this.

Sorry again I didn't respond faster to the PR.

**Comment by wbhart** [comment]
We are getting a message

Warning: ignoring unknown pragma: #pragma omp threadprivate

It seems that even recent gcc doesn't know this pragma. Is there some other way we can do this?

**Comment by wbhart** [comment]
I think configure is not setting WANT_OPENMP in config.h and there are no #if WANT_OPENMP guards around the pragmas.

**Comment by dsroche** [comment]
Hi Bill!

First, regarding the pragma warning, the way I *tried* to set it up in the
configure script, CFLAGS should either contain -fopenmp (to actually use
OpenMP, hence no warning), or -Wno-unknown-pragmas (to tell gcc to suppress
the warning message). I've tested this both ways and it works on my gcc
(6.2.0). I thought it would be much more tedious to wrap every #pragma omp
in an if/else clause. To me personally, that's the big benefit of using the
pragmas; you can just add them to the code, do absolutely nothing else, and
you get the parallelism iff OpenMP is actually enabled.

Second, regarding WANT_OPENMP, I decided that there was no need to add this
flag to the config.h, because there already is such a flag: The OpenMP
standard states that a macro _OPENMP will be defined iff OpenMP is actually
enabled. So WANT_OPENMP would be redundant and less-standard than just
using #ifdef _OPENMP when we want or need that kind of behavior. Note there
are a few such clauses in flint.h and thread_support.c.

In short, I'm surprised you're getting this warning message. it makes me
think that something went wrong with the configure or make process. Or
maybe it's something that happens on a different platform or gcc than what
I'm using. Can you send me some more details or a log so I can try to
recreate it?

Thanks,
Dan

**Comment by wbhart** [comment]
It's possible something went wrong when I configured. I'll check into it.

**Comment by wbhart** [comment]
Indeed, I hadn't run configure again on that particular machine, hence I
was missing the -Wno-unknown-pragmas. Everything is fine now.
